app/workers/gps_report_worker.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+class InvalidSupplierError < StandardError; end
+
+class GPSReportWorker
+  include Sidekiq::Worker
+  include ActionView::Helpers::TextHelper
+
+  def perform
+    @date_range = Date.yesterday - 6.days..Date.yesterday
+    threads = suppliers.map { |supplier| Thread.new { GPSReport.new(@date_range, supplier).generate } }
+    post_results(threads.each(&:join).map(&:value))
+  end
+
+private
+
+  def suppliers
+    %w[geoamey serco]
+  end
+
+  def post_results(results)
+    formatted_date_range = "#{@date_range.first.strftime('%Y/%m/%d')} - #{@date_range.last.strftime('%Y/%m/%d')}"
+    payload = {
+      blocks: [
+        {
+          type: 'header',
+          text: { type: 'plain_text', text: ":memo: GPS Data Report for #{formatted_date_range}", emoji: true },
+        },
+        { type: 'divider' },
+        *results.each_with_index.map { |res, i| generate_block(suppliers[i], res) },
+      ],
+      channel: ENV['SLACK_CHANNEL'],
+    }
+
+    slack_client.chat_postMessage(payload)
+    results.each_with_index.map { |res, i| post_failure_file(suppliers[i], res) }
+  end
+
+  def post_failure_file(supplier, results)
+    failures = results[:failures]
+    return if failures.empty?
+
+    reasons = failures.map { |r| r[:reason] }.uniq
+    failure_rows = []
+    reasons.each_with_index do |reason, reason_i|
+      failures.filter { |f| f[:reason] == reason }.each_with_index do |failure, failure_i|
+        failure_rows[failure_i] = [] if failure_rows[failure_i].nil?
+
+        failure_rows[failure_i][reason_i] = failure[:move].id
+      end
+    end
```

Comment:
```suggestion
    failure_rows = reasons.with_index.each_with_object([]) do |(reason, reason_i), failure_rows|
      failures.filter { |f| f[:reason] == reason }.each_with_index do |failure, failure_i|
        failure_rows[failure_i] = [] if failure_rows[failure_i].nil?

        failure_rows[failure_i][reason_i] = failure[:move].id
      end
    end
```

I think using `each_with_object` here is slightly neater!